### PR TITLE
Implement comment replies loading ability

### PR DESCRIPTION
**Pull request by EVAST9919**
Depends on:
- [ ] https://github.com/ppy/osu/pull/6722
- [x] https://github.com/ppy/osu/pull/6529

Implements:
- "show more" button
- "load replies" button

### CODE REVIEW

File: osu.Game/Online/API/Requests/Responses/CommentBundle.cs
Comment (bdach) [inline osu.Game/Online/API/Requests/Responses/CommentBundle.cs]: I'm confused - what is `IncludedComments` and how does it differ from `Comments`? Why does `IncludedComments`'s getter modify `Comments` too?

File: osu.Game/Online/API/Requests/Responses/CommentBundle.cs
Comment (bdach) [inline osu.Game/Online/API/Requests/Responses/CommentBundle.cs]: This snippet shows up thrice in this file, possibly factorable into a method.

File: osu.Game/Online/API/Requests/Responses/CommentBundle.cs
Comment (bdach) [inline osu.Game/Online/API/Requests/Responses/CommentBundle.cs]: Delegate duplicated twice.

File: osu.Game/Online/API/Requests/Responses/CommentBundle.cs
Comment (bdach) [inline osu.Game/Online/API/Requests/Responses/CommentBundle.cs]: Delegate duplicated twice.

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/DrawableComment.cs]: How does `childComments` differ from `loadedChildren`?

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/DrawableComment.cs]: Possibly nitpicky again, but this `Contains` on a list makes this quadratic. If possible I'd want to see `loadedChildren` being a `Dictionary` or `HashSet`, unless the order of elements matters.

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/DrawableComment.cs]: First off, this check can be simplified to
```csharp
repliesCount == 0 || children.NewValue.Count == 0
```
but mostly I'm confused with what this does. Does `children` contain only the comments that are going to be *added* below (therefore meaning that there's nothing more to load, so the button can be hidden)? I don't see any other reason for hiding the button if it's empty; if so, mind changing the parameter naming to reflect that?

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]: * Why is the parent id checked here? I assume web should take care of that, since it's passed in the request.
* `Count` is slightly inefficient here, since it won't stop at the first matching result. Use negated `.Any()` or `.All()` instead.

File: osu.Game/Overlays/Comments/ShowMoreRepliesButton.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/ShowMoreRepliesButton.cs]: I don't understand the first condition here. I assume the second is to cover the case where all the replies are loaded at once, but what's with the first?

Also in general I think this is too complicated - as far as I understand you *know* how many replies are expected in total, so if you can access the *total* number of children at the point of calling this, this should reduce to checking if the counts are equal.

**Review by bdach — CHANGES_REQUESTED** [review]
Notes from a read-through of the source alone. Haven't ran the code. I assume the tests from the previous PR will also cover this, since they're fetching online data.

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: `ChildComments` is being modified on every `GetCommentReplies` request. However, the response of this request can include already loaded comments from the main request. so `losdedChildren` is used just to check, whether the comment already exists

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: Order of the elements matters because of the `Sort` usage. We are receiving already ordered comments

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]: I could just use `Contains` but this is not working because some data can be differ for one comment but different requests (for example, in one request comment don't have children, but in another it has)

File: osu.Game/Overlays/Comments/ShowMoreRepliesButton.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/ShowMoreRepliesButton.cs]: First condition is for the case, when we know, what there are replies exist, but they are no loaded and atm we don't have any children (in this case `LoadMoreCommentsButton` will be shown)

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]: We can receive the same comment twice (first time in the main request, second - in the `GetRepliesRequest`)

File: osu.Game/Online/API/Requests/Responses/CommentBundle.cs
Comment (EVAST9919) [reply osu.Game/Online/API/Requests/Responses/CommentBundle.cs]: `IncludedComments` contains all non-top-level comments (in case of `GetRepliesRequest`)

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (bdach) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: I'd either advocate for a naming change or an xmldoc here for clarity (ie. `requestedComments` vs `loadedComments`?)

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (bdach) [inline osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]:
```suggestion
                if (children.All(child => child.Id != c.Id))
```
Clears up a resharper warning visible in the CI build log.

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: Removed by using `children.OldValue` in `childComments.BindValueChanged()`

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]: Is this not guaranteed by the request? If it is I'm thinking this can be refactored further by using `BindableList<T>`.

In here:
```
// perform request
foreach (var comment in response)
    list.Add(comment)
```

In DrawableComment:
```
list.OnAdded += c =>  /* add new drawable comment */
```

which would remove the whole bindvaluechanged + foreach stuff.

File: osu.Game/Overlays/Comments/GetCommentRepliesButton.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/GetCommentRepliesButton.cs]: I'm not really a fan of this, modifying `ChildComments` from a button. I'd much rather have this button expose an event like `Action<List<Comment>> OnCommentsReceived`

File: osu.Game/Overlays/Comments/LoadRepliesButton.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/LoadRepliesButton.cs]: Relating to my previous comment re: exposing an event instead... This isn't the correct way to do this anyway - shouldn't be changing your own visual state.

File: osu.Game/Overlays/Comments/ShowChildrenButton.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/ShowChildrenButton.cs]: Same here, this is also not the proper way to do this (mutating your own visual state). It's the responsibility of the user of this button to do so.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: Resolved by latest refactoring

**Review by bdach — DISMISSED** [review]
Source definitely looking better.

Ran the code, and unfortunately some Markdown links don't seem to be working correctly:
image
The format seems a little bogus, might be potentially a web-side issue?

File: osu.Game/Overlays/Comments/DrawableComment.cs
Comment (EVAST9919) [reply osu.Game/Overlays/Comments/DrawableComment.cs]: oops

**Comment by EVAST9919** [comment]
> Ran the code, and unfortunately some Markdown links don't seem to be working correctly:
> image
> The format seems a little bogus, might be potentially a web-side issue?

This is how message looks like
`[http://puu.sh/a9Clg/143ed6c...](http://puu.sh/a9Clg/143ed6c903.png "http://puu.sh/a9Clg/143ed6c903.png") ..... D:`

**Comment by bdach** [comment]
I checked one of the other steps (the one with the changelog) and the link worked properly. I assume the above example is some vestige of a partially-broken conversion.

A bit tentative on approving since the others might have objections, but looks alright to me - dismissing review.

**Comment by bdach** [comment]
Just as a side note, I have a draft of a fix for the markdown formatting, but it needs some work still and I don't want to rush it. Expect it within a day or two, though.

File: osu.Game/Overlays/Comments/CommentsContainer.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/CommentsContainer.cs]: Methods shouldn't be above ctor

File: osu.Game/Overlays/Comments/CommentsContainer.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/CommentsContainer.cs]: Maybe rename to `noCommentsPlaceholder`

File: osu.Game/Overlays/Comments/CommentsContainer.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Comments/CommentsContainer.cs]: How about this being the method that clears and retrieves comments, and having the sort change event trigger `ShowComments(type, id)`?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by EVAST9919** [comment]
Will be reopened after some refactoring and "offline-mode" implementation

**Comment by bdach** [comment]
I've got to admit I'm a little bit miffed this has grown in scope again. Your changes *need* to be more incremental, or else it's hard to review things properly.

I'll take a look at this later, but please split things more for the good of everyone in the future (and that includes you - smaller PRs = faster merging).

**Comment by EVAST9919** [comment]
Will be reopened after dependency merge

**Pull request closed without merging**
